compiler/resolution/visibleFunctions.cpp

```diff
@@ -548,13 +655,132 @@ static void getVisibleMethodsFromUseList(const char* name, CallExpr* call,
+                                    visibleFns, true, namedTypes);
+        } else {
+          getVisibleMethodsImpl(name, call, mod->block, visInfo, visited,
+                                visibleFns, true);
+        }
       }
     }
   }
 }
 
+// This function returns a set containing the intersection of outer and inner.
+// Any symbol that is solely in one list and not the other is not valid and
+// should not be returned.
+static void mergeFilters(std::set<const char*> outer,
+                         std::set<const char*> inner,
+                         std::set<const char*>* merged) {
+  INT_ASSERT(outer.size() > 0 && inner.size() > 0);
+  std::set<const char*>::iterator it;
+
+  for (it = outer.begin(); it != outer.end(); ++it) {
+    if (inner.find(*it) != inner.end()) {
+      merged->insert(*it);
+    }
+  }
+}
+
+static void getVisibleMethodsFiltered(const char* name, CallExpr* call,
+                                      BlockStmt* block, VisibilityInfo* visInfo,
+                                      std::set<BlockStmt*>& visited,
+                                      Vec<FnSymbol*>& visibleFns,
+                                      bool inUseChain,
+                                      std::set<const char*> typeNames) {
+  if (block == rootBlock) return; // nothing there
+  //
+  // avoid infinite recursion due to modules with mutual uses
+  //
+  const bool firstVisit = (visited.find(block) == visited.end());
+  if (!firstVisit && inUseChain) {
+    // inUseChain tracks how we are currently finding the scope.  There's
+    // three scenarios:
+    // - found the scope the first time by traversing a use/import
+    //   - found the scope again by also traversing a use/import (which causes
+    //      us to enter this if branch)
+    //   - found the scope again by going up in scope from the call site (this
+    //     case will *not* cause us to enter this if branch)
+    // - found the scope the first time by going up in scope from the call site
+    //   and again by traversing a use/import (this case will also cause us to
+    //   enter this if branch)
+
+    // We don't want to traverse private use/import statements unless we found
+    // the scope with the use/import statement by going up in scope from the
+    // call site.
+    return;
+  }
+
+  BlockStmt* instantiationPt = getVisibleFnsInstantiationPt(block);
+
+  // print tracking information
+  if (call->id == breakOnResolveID) {
+    getVisibleFnsShowBlock("methods", block, instantiationPt);
+  }
+
+  if (firstVisit)
+    getVisibleMethodsFirstVisitFiltered(name, call, block, visInfo, visited,
+                                        visibleFns, typeNames);
+
+  if (block->useList != NULL) {
+    getVisibleMethodsFromUseListFiltered(name, call, block, visInfo, visited,
+                                         visibleFns, inUseChain, firstVisit,
+                                         typeNames);
+  }
+
+  // TODO: double check this is right
```

Comment:
is there something to do here?

Reply:
I'm not entirely convinced the code is necessary, but I was worried about not adding it.  It's a holdover from the function this was based on.

Reply:
I'll remove the TODO at least